src/cockpit/channels/http_stream2.py

```diff
@@ -0,0 +1,130 @@
+            logger.error("Failed to open %s:%s [connection: %s]: %s", address, port, connection, e)
+            self.close(problem='not-found', message=str(e))
+            return None
+
+        if connection:
+            HttpStream2Channel.connections[connection] = con
+        return con
+
+    def read_send_response(self, response):
+        '''Completely read the response and send it to the channel'''
+
+        while True:
+            # we want to stream data blocks as soon as they come in
+            block = response.read1(4096)
+            if not block:
+                logger.debug('reading response done')
+                # this returns immediately and does not read anything more, but updates the http.client's
+                # internal state machine to "response done"
+                block = response.read()
+                assert block == b''
+                break
+            logger.debug('read block of size %i', len(block))
+            self.send_data(block)
+
+    def parse_headers(self, http_msg):
+        headers = dict(http_msg)
+        remove = ["Connection", "Transfer-Encoding"]
+        if self.is_binary:
+            remove = ["Content-Length", "Range"]
+        for h in remove:
+            try:
+                del headers[h]
+            except KeyError:
+                pass
+        return headers
+
+    def request(self):
+        if not self.conn:
+            self.conn = self.create_connection()
+            if not self.conn:
+                # make_connection does the error reporting
+                return
+
+        self.conn.request(self.options.get('method'), self.options.get('path'), headers=self.options.get('headers') or {})
+        response = self.conn.getresponse()
+        self.send_control(command="response",
+                          status=response.status,
+                          reason=response.reason,
+                          headers=self.parse_headers(response.headers))
+        self.read_send_response(response)
+
+        self.done()
+        self.close()
+        logger.debug('closed')
+
+    def do_open(self, options):
+        logger.debug('open %s', options)
+        self.is_binary = options.get('binary')
+        if not options.get('method'):
+            self.close(problem='protocol-error', message='missing "method" field in HTTP stream request')
+            return
+        if not options.get('path'):
+            self.close(problem='protocol-error', message='missing "path" field in HTTP stream request')
+            return
+
+        self.conn = self.get_connection(options.get('connection'))
+        if not self.conn:
+            # these are not required/ignored for shared connections
+            port = options.get('port')
+            unix = options.get('unix')
+            if not port and not unix:
+                self.close(problem='protocol-error', message='no "port" or "unix" option for channel')
+                return
+            if port and unix:
```

Comment:
`if port is not None and unix is not None`

etc etc :)

Reply:
Done.